clear-metrics-descriptors: docs and usability

johnsonj opened the pull request:
- Added usability improvements to the clear-metrics-descriptors script.
  This adds a confirmation prompt, presence validation for the
  GCP_PROJECT_ID, and cleaned up output.
- Add specific instructions for clearing out metrics descriptors. This
  doc omits the need to use the develop branch. That's fine- if someone
  does follow these steps today it will work.
- Add general 2.0 upgrade instructions with guidance on singleton
  instance of the nozzle and the need for clear-metrics-descriptors.

updates #185 

/cc @knyar @fluffle

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cloudfoundry-community/stackdriver-tools/198)
<!-- Reviewable:end -->

knyar commented:
Nice! I've checked the cleanup script on `cre-pcf` and it seems to work fine.

---

Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[docs/upgrading_2.0.md, line 22 at r1](https://reviewable.io:443/reviews/cloudfoundry-community/stackdriver-tools/198#-L59VlNiQxYNMKQnJVnp:-L59VlNiQxYNMKQnJVnq:bsbejv1) ([raw file](https://github.com/cloudfoundry-community/stackdriver-tools/blob/f720c0343fe28926cdd58f8be7f8be08b5c89f27/docs/upgrading_2.0.md#L22)):*
> ```Markdown
> This issue is mitigated by running a single instance of the nozzle. 
> An upstream feature in Loggregator is being [tracked](https://www.pivotaltracker.com/n/projects/993188/stories/154821450) to fix this.
> It is highly recommended that users evaluate running a single instance. This release significantly improves
> ```

Perhaps we should more strongly state that the tile actually forces users to run a single instance now? Something like this:

The tile currently forces the nozzle job to have a single instance, and users are encouraged to vertically scale the nozzle VM in case of resource saturation. A number of performance optimizations in this release make sure that a single instance of the nozzle is able to provide enough throughput for medium sized PCF installations.

If you find the need to scale the nozzle job horizontally, you will need to adjust nozzle bosh job parameters, instead of relying on nozzle being configured via the Ops Manager tile. There are two possible ways forward:
- if the nozzle is heavily used to proxy log events, you can run two separate nozzle bosh jobs: one for log messages (which can be scaled horizontally), and the other for metrics (which will need to have a single instance). Nozzle jobs can be configured to receive only a subset of events via `firehose.events_to_stackdriver_logging` and `firehose.events_to_stackdriver_monitoring` bosh manifest properties.
- you can disable cumulative counter tracking completely by setting `nozzle.enable_cumulative_counters` to false. This will revert to the old behavior of exporting each counter as a pair of gauges (with `.delta` and `.total` suffixes). With cumulative counters disabled, a single nozzle job can be configured to run with multiple instances.

---

*[src/stackdriver-nozzle/docs/clear-metrics-descriptors.md, line 20 at r1](https://reviewable.io:443/reviews/cloudfoundry-community/stackdriver-tools/198#-L59ZI8hU-_k6v9CkPsG:-L59ZI8hU-_k6v9CkPsH:b-yj7cj9) ([raw file](https://github.com/cloudfoundry-community/stackdriver-tools/blob/f720c0343fe28926cdd58f8be7f8be08b5c89f27/src/stackdriver-nozzle/docs/clear-metrics-descriptors.md#L20)):*
> ```Markdown
> ```
> 
> ## 2. Fetch and build `clear-metrics-descriptors
> ```

The second backtick seems to be missing.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cloudfoundry-community/stackdriver-tools/198#-:-L59_yp0YXE9HAtu_INL:bfxxr5h)*
<!-- Sent from Reviewable.io -->

fluffle commented:
Review status: 0 of 3 files reviewed at latest revision, 5 unresolved discussions.

---

*[src/stackdriver-nozzle/cmd/clear-metrics-descriptors.go, line 51 at r1](https://reviewable.io:443/reviews/cloudfoundry-community/stackdriver-tools/198#-L59cFMBq_aH1c3aokre:-L59cFMBq_aH1c3aokrf:b-9446k8) ([raw file](https://github.com/cloudfoundry-community/stackdriver-tools/blob/f720c0343fe28926cdd58f8be7f8be08b5c89f27/src/stackdriver-nozzle/cmd/clear-metrics-descriptors.go#L51)):*
> ```Go
> )
> 
> const projectIdEnv = "GCP_PROJECT_ID"
> ```

I know that env vars are the way configuration gets into the nozzle, but we should definitely provide a flag for this as well since this is a standalone CLI app.

While we're at it, a second flag that takes a regexp to match against the custom metric name would be great, so it is possible for people to selectively remove metric descriptors. (I've been meaning to send a patch to do this for ages, but i've not needed that functionality recently.)

---

*[src/stackdriver-nozzle/cmd/clear-metrics-descriptors.go, line 75 at r1](https://reviewable.io:443/reviews/cloudfoundry-community/stackdriver-tools/198#-L59g5WP6Y0Qye2Hs-zD:-L59g5WP6Y0Qye2Hs-zE:bwdmtmx) ([raw file](https://github.com/cloudfoundry-community/stackdriver-tools/blob/f720c0343fe28926cdd58f8be7f8be08b5c89f27/src/stackdriver-nozzle/cmd/clear-metrics-descriptors.go#L75)):*
> ```Go
> 	for {
> 		resp, err := it.Next()
> 		if err == iterator.Done {
> ```

Any reason you didn't do:

```golang
for resp, err := it.Next(); err != iterator.Done; resp, err = it.Next() {
    // ...
}
```

instead of these first four lines?

---

*[src/stackdriver-nozzle/docs/clear-metrics-descriptors.md, line 23 at r1](https://reviewable.io:443/reviews/cloudfoundry-community/stackdriver-tools/198#-L59hbwjcFW5s6M8iK8C:-L59hbwjcFW5s6M8iK8D:b-stej3d) ([raw file](https://github.com/cloudfoundry-community/stackdriver-tools/blob/f720c0343fe28926cdd58f8be7f8be08b5c89f27/src/stackdriver-nozzle/docs/clear-metrics-descriptors.md#L23)):*
> ```Markdown
> ```bash
> go get -d github.com/cloudfoundry-community/stackdriver-tools/src/stackdriver-nozzle
> GOBIN=`pwd` go install $(go env GOPATH)/src/github.com/cloudfoundry-community/stackdriver-tools/src/stackdriver-nozzle/cmd/clear-metrics-descriptors.go
> ```

It's odd that you provide different instructions here than you do in the comments at the top of clear-metric-descriptors.go. The `go run` approach works fine, why do people need to jump through these extra hoops?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cloudfoundry-community/stackdriver-tools/198)*
<!-- Sent from Reviewable.io -->

johnsonj commented:
Review status: 0 of 3 files reviewed at latest revision, 5 unresolved discussions.

---

*[docs/upgrading_2.0.md, line 22 at r1](https://reviewable.io:443/reviews/cloudfoundry-community/stackdriver-tools/198#-L59VlNiQxYNMKQnJVnp:-L5ABmt_gRGWJFLK_T-3:b-v1cy3s) ([raw file](https://github.com/cloudfoundry-community/stackdriver-tools/blob/f720c0343fe28926cdd58f8be7f8be08b5c89f27/docs/upgrading_2.0.md#L22)):*
<details><summary><i>Previously, knyar (Anton Tolchanov) wrote…</i></summary><blockquote>

Perhaps we should more strongly state that the tile actually forces users to run a single instance now? Something like this:

The tile currently forces the nozzle job to have a single instance, and users are encouraged to vertically scale the nozzle VM in case of resource saturation. A number of performance optimizations in this release make sure that a single instance of the nozzle is able to provide enough throughput for medium sized PCF installations.

If you find the need to scale the nozzle job horizontally, you will need to adjust nozzle bosh job parameters, instead of relying on nozzle being configured via the Ops Manager tile. There are two possible ways forward:
- if the nozzle is heavily used to proxy log events, you can run two separate nozzle bosh jobs: one for log messages (which can be scaled horizontally), and the other for metrics (which will need to have a single instance). Nozzle jobs can be configured to receive only a subset of events via `firehose.events_to_stackdriver_logging` and `firehose.events_to_stackdriver_monitoring` bosh manifest properties.
- you can disable cumulative counter tracking completely by setting `nozzle.enable_cumulative_counters` to false. This will revert to the old behavior of exporting each counter as a pair of gauges (with `.delta` and `.total` suffixes). With cumulative counters disabled, a single nozzle job can be configured to run with multiple instances.
</blockquote></details>

Agreed!  I've worked in your suggestion. PTAL

---

*[src/stackdriver-nozzle/cmd/clear-metrics-descriptors.go, line 51 at r1](https://reviewable.io:443/reviews/cloudfoundry-community/stackdriver-tools/198#-L59cFMBq_aH1c3aokre:-L5AHrbvLeoREslFmV4B:b-n431gr) ([raw file](https://github.com/cloudfoundry-community/stackdriver-tools/blob/f720c0343fe28926cdd58f8be7f8be08b5c89f27/src/stackdriver-nozzle/cmd/clear-metrics-descriptors.go#L51)):*
<details><summary><i>Previously, fluffle (Alex Bee) wrote…</i></summary><blockquote>

I know that env vars are the way configuration gets into the nozzle, but we should definitely provide a flag for this as well since this is a standalone CLI app.

While we're at it, a second flag that takes a regexp to match against the custom metric name would be great, so it is possible for people to selectively remove metric descriptors. (I've been meaning to send a patch to do this for ages, but i've not needed that functionality recently.)


</blockquote></details>

Moved to flag. I went the lazy route and made the prefix configurable. AFAIK that takes care of most scenarios I can think of.

---

*[src/stackdriver-nozzle/cmd/clear-metrics-descriptors.go, line 75 at r1](https://reviewable.io:443/reviews/cloudfoundry-community/stackdriver-tools/198#-L59g5WP6Y0Qye2Hs-zD:-L5ABxyDCtY9vgzaIWrB:b-cyabct) ([raw file](https://github.com/cloudfoundry-community/stackdriver-tools/blob/f720c0343fe28926cdd58f8be7f8be08b5c89f27/src/stackdriver-nozzle/cmd/clear-metrics-descriptors.go#L75)):*
<details><summary><i>Previously, fluffle (Alex Bee) wrote…</i></summary><blockquote>

Any reason you didn't do:

```golang
for resp, err := it.Next(); err != iterator.Done; resp, err = it.Next() {
    // ...
}
```

instead of these first four lines?
</blockquote></details>

I didn't write this- but no reason AFAIK. I'll clean it up. 

---

*[src/stackdriver-nozzle/docs/clear-metrics-descriptors.md, line 23 at r1](https://reviewable.io:443/reviews/cloudfoundry-community/stackdriver-tools/198#-L59hbwjcFW5s6M8iK8C:-L5AD8FxHMrLfUrCRxAp:bky1x2f) ([raw file](https://github.com/cloudfoundry-community/stackdriver-tools/blob/f720c0343fe28926cdd58f8be7f8be08b5c89f27/src/stackdriver-nozzle/docs/clear-metrics-descriptors.md#L23)):*
<details><summary><i>Previously, fluffle (Alex Bee) wrote…</i></summary><blockquote>

It's odd that you provide different instructions here than you do in the comments at the top of clear-metric-descriptors.go. The `go run` approach works fine, why do people need to jump through these extra hoops?
</blockquote></details>

I was thinking that i'd be nice to ensure they're build environment is properly setup before doing the nozzle tear down step. On further review it's not necessary. Updating to go run. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cloudfoundry-community/stackdriver-tools/198)*
<!-- Sent from Reviewable.io -->

fluffle commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Thanks!

---

Review status: 0 of 3 files reviewed at latest revision, 7 unresolved discussions.

---

*[docs/upgrading_2.0.md, line 5 at r2](https://reviewable.io:443/reviews/cloudfoundry-community/stackdriver-tools/198#-L5DqTrXy_yp8gKS6Ecw:-L5DqTrXy_yp8gKS6Ecx:b-pxyjek) ([raw file](https://github.com/cloudfoundry-community/stackdriver-tools/blob/6195dae71b8b6effc0cc92ab94008f34aa90abbd/docs/upgrading_2.0.md#L5)):*
> ```Markdown
> This release of `stackdriver-tools` introduces significant breaking changes for `stackdriver-nozzle` users.
> These breaking changes provide improved metrics functionality but require the operator to reset 
> aspects of their project. This will result in irreversible historical data loss and the need to 
> ```

I think there's a work-around for this, but i'm not sure if it's worth mentioning. If users really want to preserve history, they can create a new Stackdriver project (with its own 500-descriptor limit) for the new Nozzle to write its data to.

---

*[docs/upgrading_2.0.md, line 8 at r2](https://reviewable.io:443/reviews/cloudfoundry-community/stackdriver-tools/198#-L5Dr7Hmh0JmGdz7_dfl:-L5Dr7Hmh0JmGdz7_dfm:bd1g3kd) ([raw file](https://github.com/cloudfoundry-community/stackdriver-tools/blob/6195dae71b8b6effc0cc92ab94008f34aa90abbd/docs/upgrading_2.0.md#L8)):*
> ```Markdown
> recreate dashboards.
> 
> This release of the `stackdriver-nozzle` changes the name, type, and labels of all reported metrics.
> ```

Is it worth providing some details of the changes here, so people can understand them? Broadly:

Metric changes:
*   Metrics derived from the firehose now have the metric origin prefixed to the metric name, e.g. "gorouter.total_requests".
*   Metrics derived from the firehose are now exported under a configurable path prefix, which defaults to "firehose".
*   CounterEvent metrics are now represented as COUNTERs rather than GAUGEs in Stackdriver, so per-second rates can be computed.
*   The nozzle exports its own custom metrics under the "stackdriver-nozzle" prefix.

Label changes:
*   Application, space and organization UUIDs are no longer exported.
*   Resolved application, space and organization names are concatenated into a single "applicationPath" label.
*   The application instance index is now exported.
*   Event envelope tags are now exported as a string of comma-separated key=value pairs.
*   Origin and EventType labels are only attached to logs, not metrics.
*   There is a new "foundation" label that is configurable per nozzle instance, to distinguish between multiple foundations in a GCP / Stackdriver project.

---

*[src/stackdriver-nozzle/cmd/clear-metrics-descriptors.go, line 51 at r1](https://reviewable.io:443/reviews/cloudfoundry-community/stackdriver-tools/198#-L59cFMBq_aH1c3aokre:-L5Dpg1qbtYVk2uudnhc:b-amdl3a) ([raw file](https://github.com/cloudfoundry-community/stackdriver-tools/blob/f720c0343fe28926cdd58f8be7f8be08b5c89f27/src/stackdriver-nozzle/cmd/clear-metrics-descriptors.go#L51)):*
<details><summary><i>Previously, johnsonj (Jeff Johnson) wrote…</i></summary><blockquote>

Moved to flag. I went the lazy route and made the prefix configurable. AFAIK that takes care of most scenarios I can think of.
</blockquote></details>

Sure, that works for me. Prefix matching works well with the nozzle's new metric hierarchy and PCFs requirement that the metric origin is prefixed.

---

*[src/stackdriver-nozzle/cmd/clear-metrics-descriptors.go, line 75 at r1](https://reviewable.io:443/reviews/cloudfoundry-community/stackdriver-tools/198#-L59g5WP6Y0Qye2Hs-zD:-L5Dq6_58sa3hMi0ZZOr:bx23kzr) ([raw file](https://github.com/cloudfoundry-community/stackdriver-tools/blob/f720c0343fe28926cdd58f8be7f8be08b5c89f27/src/stackdriver-nozzle/cmd/clear-metrics-descriptors.go#L75)):*
<details><summary><i>Previously, johnsonj (Jeff Johnson) wrote…</i></summary><blockquote>

I didn't write this- but no reason AFAIK. I'll clean it up. 
</blockquote></details>

Ooo, sorry. I saw it in green and blindly assumed it was new code; my bad. I should read more carefully :-(

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cloudfoundry-community/stackdriver-tools/198#-:-L5Du8WOMArQNo26Rfn5:b-g2a5fb)*
<!-- Sent from Reviewable.io -->

johnsonj commented:
Review status: 0 of 3 files reviewed at latest revision, 4 unresolved discussions.

---

*[docs/upgrading_2.0.md, line 5 at r2](https://beta.reviewable.io/reviews/cloudfoundry-community/stackdriver-tools/198#-L5DqTrXy_yp8gKS6Ecw:-L5GcTVZ-VG_n7kw8EuK:b-6m8rc5) ([raw file](https://github.com/cloudfoundry-community/stackdriver-tools/blob/6195dae71b8b6effc0cc92ab94008f34aa90abbd/docs/upgrading_2.0.md#L5)):*
<details><summary><i>Previously, fluffle (Alex Bee) wrote…</i></summary><blockquote>

I think there's a work-around for this, but i'm not sure if it's worth mentioning. If users really want to preserve history, they can create a new Stackdriver project (with its own 500-descriptor limit) for the new Nozzle to write its data to.
</blockquote></details>

It's worth a mention, I've incorporated it into Clear Metric Descriptors section. 

---

*[docs/upgrading_2.0.md, line 8 at r2](https://beta.reviewable.io/reviews/cloudfoundry-community/stackdriver-tools/198#-L5Dr7Hmh0JmGdz7_dfl:-L5GcQqW7kPqom0janbs:bev550q) ([raw file](https://github.com/cloudfoundry-community/stackdriver-tools/blob/6195dae71b8b6effc0cc92ab94008f34aa90abbd/docs/upgrading_2.0.md#L8)):*
<details><summary><i>Previously, fluffle (Alex Bee) wrote…</i></summary><blockquote>

Is it worth providing some details of the changes here, so people can understand them? Broadly:

Metric changes:
*   Metrics derived from the firehose now have the metric origin prefixed to the metric name, e.g. "gorouter.total_requests".
*   Metrics derived from the firehose are now exported under a configurable path prefix, which defaults to "firehose".
*   CounterEvent metrics are now represented as COUNTERs rather than GAUGEs in Stackdriver, so per-second rates can be computed.
*   The nozzle exports its own custom metrics under the "stackdriver-nozzle" prefix.

Label changes:
*   Application, space and organization UUIDs are no longer exported.
*   Resolved application, space and organization names are concatenated into a single "applicationPath" label.
*   The application instance index is now exported.
*   Event envelope tags are now exported as a string of comma-separated key=value pairs.
*   Origin and EventType labels are only attached to logs, not metrics.
*   There is a new "foundation" label that is configurable per nozzle instance, to distinguish between multiple foundations in a GCP / Stackdriver project.




</blockquote></details>

Excellent summary. Added!

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cloudfoundry-community/stackdriver-tools/198)*
<!-- Sent from Reviewable.io -->

johnsonj commented:
Review status: 0 of 3 files reviewed at latest revision, 4 unresolved discussions.

---

*[src/stackdriver-nozzle/docs/clear-metrics-descriptors.md, line 20 at r1](https://beta.reviewable.io/reviews/cloudfoundry-community/stackdriver-tools/198#-L59ZI8hU-_k6v9CkPsG:-L5GcgOYCrBgFtl1WsTf:b-896fix) ([raw file](https://github.com/cloudfoundry-community/stackdriver-tools/blob/f720c0343fe28926cdd58f8be7f8be08b5c89f27/src/stackdriver-nozzle/docs/clear-metrics-descriptors.md#L20)):*
<details><summary><i>Previously, knyar (Anton Tolchanov) wrote…</i></summary><blockquote>

The second backtick seems to be missing.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cloudfoundry-community/stackdriver-tools/198)*
<!-- Sent from Reviewable.io -->

johnsonj commented:
Reviewed 3 of 3 files at r2.
Review status: 2 of 3 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/cloudfoundry-community/stackdriver-tools/198)*
<!-- Sent from Reviewable.io -->
